src/main/java/com/experoinc/janusgraph/diskstorage/foundationdb/FoundationDBTx.java

```diff
@@ -207,27 +206,32 @@ private TransactionClose(String msg) {
     public synchronized  Map<KVQuery, List<KeyValue>> getMultiRange(final List<Object[]> queries)
             throws PermanentBackendException {
         Map<KVQuery, List<KeyValue>> resultMap = new ConcurrentHashMap<>();
-        final List<CompletableFuture> futures = new LinkedList<>();
         try {
-            for(Object[] obj : queries) {
-                final KVQuery query = (KVQuery) obj[0];
-                final byte[] start = (byte[]) obj[1];
-                final byte[] end = (byte[]) obj[2];
-
-                futures.add(tx.getRange(start, end, query.getLimit()).asList()
-                        .whenComplete((res, th) -> {
-                            if (th == null) {
-                                if (res == null) {
-                                    res = Collections.emptyList();
-                                }
-                                resultMap.put(query, res);
-                            } else {
-                                log.error("failed to complete getRange of multiRange", th);
-                                throw new RuntimeException("failed to getRange for multiRangeQuery", th);
-                            }
-                        }));
+            for (int i = 0; i < maxRuns; i++) {
+                final List<CompletableFuture> futures = new LinkedList<>();
+                for (Object[] obj: queries) {
+                    final KVQuery query = (KVQuery) obj[0];
+                    final byte[] start = (byte[]) obj[1];
+                    final byte[] end = (byte[]) obj[2];
+                    if(!resultMap.containsKey(query)) {
```

Comment:
if all futures of the current retry pass, then resultMap will have the result of the current getRange query and hence rest all loops will complete without triggering any extra completableFuture.  yes, a few cpu cycles will be waster but this makes code a bit short and more readable